cmd/clusterctl/client/repository/repository_gitlab_test.go

```diff
@@ -0,0 +1,206 @@
+				variableClient: test.NewFakeVariableClient(),
+			},
+			want:      nil,
+			wantedErr: "invalid url: a GitLab repository url should be in the form https://{host}/api/v4/projects/{projectSlug}/packages/generic/{packageName}/{defaultVersion}/{componentsPath}",
+		},
+		{
+			name: "provider url should have the correct number of parts",
+			field: field{
+				providerConfig: config.NewProvider("test", "https://gitlab.example.org/api/v4/projects/group%2Fproject/packages/generic/my-package/v1.0/path/grrr", clusterctlv1.CoreProviderType),
+				variableClient: test.NewFakeVariableClient(),
+			},
+			want:      &gitLabRepository{},
+			wantedErr: "invalid url: a GitLab repository url should be in the form https://{host}/api/v4/projects/{projectSlug}/packages/generic/{packageName}/{defaultVersion}/{componentsPath}",
+		},
+		{
+			name: "provider url should have the correct prefix",
+			field: field{
+				providerConfig: config.NewProvider("test", "https://gitlab.example.org/subpath/api/v4/projects/group%2Fproject/packages/generic/my-package/v1.0/path", clusterctlv1.CoreProviderType),
+				variableClient: test.NewFakeVariableClient(),
+			},
+			want:      &gitLabRepository{},
+			wantedErr: "invalid url: a GitLab repository url should be in the form https://{host}/api/v4/projects/{projectSlug}/packages/generic/{packageName}/{defaultVersion}/{componentsPath}",
+		},
+		{
+			name: "provider url should have the packages part",
+			field: field{
+				providerConfig: config.NewProvider("test", "https://gitlab.example.org/api/v4/projects/group%2Fproject/packages-invalid/generic/my-package/v1.0/path", clusterctlv1.CoreProviderType),
+				variableClient: test.NewFakeVariableClient(),
+			},
+			want:      nil,
+			wantedErr: "invalid url: a GitLab repository url should be in the form https://{host}/api/v4/projects/{projectSlug}/packages/generic/{packageName}/{defaultVersion}/{componentsPath}",
+		},
+		{
+			name: "provider url should have the generic part",
+			field: field{
+				providerConfig: config.NewProvider("test", "https://gitlab.example.org/api/v4/projects/group%2Fproject/packages/generic-invalid/my-package/v1.0/path", clusterctlv1.CoreProviderType),
+				variableClient: test.NewFakeVariableClient(),
+			},
+			want:      nil,
+			wantedErr: "invalid url: a GitLab repository url should be in the form https://{host}/api/v4/projects/{projectSlug}/packages/generic/{packageName}/{defaultVersion}/{componentsPath}",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			resetCaches()
+
+			gitLab, err := NewGitLabRepository(tt.field.providerConfig, tt.field.variableClient)
+			if tt.wantedErr != "" {
+				g.Expect(err).To(MatchError(tt.wantedErr))
+				return
+			}
+
+			g.Expect(err).NotTo(HaveOccurred())
+			g.Expect(gitLab).To(Equal(tt.want))
+		})
+	}
+}
+
+func Test_gitLabRepository_getFile(t *testing.T) {
+	client, mux, serverURL, teardown := test.NewFakeHTTPsServer()
```

Comment:
I don't know if it is worth to have a separate helper function to wrap httptest.